### PR TITLE
Add tests for active-repositories feature

### PR DESCRIPTION
**Pull request by erikd**
This PR only contains tests, which were generated by Claude code, and then manually reviewed. Unfortunately I have not worked much on Cabal so my manual review may not be up-to-scratch.

This PR was submitted in preparation for taking over the work in https://github.com/haskell/cabal/pull/8997 and getting it merged. One of the things mentioned in that PR was that need for tests for the `active-repositories` feature.

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](CONTRIBUTING.md#other-conventions).
* [ ] Is this a PR that fixes CI? If so, it will need to be backported to older cabal release branches (ask maintainers for directions).

### CODE REVIEW

**Comment by erikd** [comment]
I asked Claude to give me an idea of the test coverage. Its response:

> The top commit (6898e42c571b77e4c2e0b0b378eab209fe91e6e9) captures the IndexUtils/ActiveRepos.hs and the index-combining 
> tests we added to IndexUtils.hs.
>
>  So the breakdown for what this commit actually adds vs the total feature:                                                            
>
> | Layer | Coverage |
> |---------|--------------------------------------|
>  | Layer 1| ~80%|
>  | Layer 2|  ~65%|    
>  |Layer 3|  ~58%|    
> |Layer 4|  ~30%|
> |Layer 5| ~0%|

Where:
  * Layer 1 — pure logic (ActiveRepos.hs)
  * Layer 2 — project config parsing (pre-existing tests only, nothing new in this PR)
  * Layer 3 — index combining (IndexUtils.hs) (all 14 combining tests are in this PR)
  * Layer 4 — freeze file (filterSkippedActiveRepos unit tests in this PR)
  * Layer 5 — solver/planning wiring

File: cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs [hunk 1]
```diff
@@ -80,3 +88,115 @@ simpleVersionsParserTests =
...
+          @?= sort [foo1, bar1]
+    , testCase "Merge+Merge: different versions of same package are both visible" $
+        pkgs [(repoFoo1, CombineStrategyMerge), (repoFoo2, CombineStrategyMerge)]
+          @?= sort [foo1, foo2]
+    , testCase "Merge+Override: packages only in first repo remain visible" $
+        pkgs [(repoFoo1, CombineStrategyMerge), (repoBar1, CombineStrategyOverride)]
+          @?= sort [foo1, bar1]
+    , testCase "Merge+Override: override repo replaces all versions of overlapping package" $
+        -- repoFoo12 has foo-1.0 and foo-1.1; repoFoo2 has only foo-2.0.
+        -- Override means repoFoo2 wins the entire 'foo' bucket.
```
Comment (andreabedini) [inline cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs:128]: This comments seems wrong. Below we have
```
repoFoo12 = PackageIndex.fromList [foo1, foo2]
```

So `repoFoo12` has `foo-1.0` and `foo-2.0`. I am not sure what was intended.

File: cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs [hunk 2]
```diff
@@ -80,3 +88,115 @@ simpleVersionsParserTests =
...
+  -> PackageIndex.PackageIndex PackageIdentifier
+combineIndex acc (_, CombineStrategySkip) = acc
+combineIndex acc (idx, CombineStrategyMerge) = PackageIndex.merge acc idx
+combineIndex acc (idx, CombineStrategyOverride) = PackageIndex.override acc idx
+
+-- Run the combining fold and return the result as a sorted list of PackageIds.
+pkgs
+  :: [(PackageIndex.PackageIndex PackageIdentifier, CombineStrategy)]
+  -> [PackageIdentifier]
+pkgs = sort . PackageIndex.allPackages . foldl combineIndex mempty
```
Comment (andreabedini) [inline cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs]:
```suggestion
pkgs = sort . PackageIndex.allPackages . foldl' combineIndex mempty
```

This is extra picky but since we say `Mirrors the addIndex / foldl' in getSourcePackagesAtIndexState` we might as well reproduce the same behaviour.

BTH I am not sure testing a reimplementation is a good idea, it will go out of sync with the actual code and lose all its value. Can we perhaps export from `IndexUtils` the functions we want to test here?

**Review by andreabedini — APPROVED** [review]
I left a couple of comments but otherwise LGTM.

Not a deal breaker but any chance you can add a simple test for [this edge case behaviour](https://github.com/erikd/cabal/blob/335c948beb59b10bed88662e161b93bec6eb4aca/cabal-install/src/Distribution/Client/IndexUtils/ActiveRepos.hs#L164-L165)?
```
-- Note: currently if 'ActiveRepoRest' is provided more than once,
-- rest-repositories will be multiple times in the output.
```

**Comment by andreabedini** [comment]
A changelog file is missing, here is a suggestion:
```
---
synopsis: Add unit tests for active-repositories feature
packages: [cabal-install]
prs: 11684
---

Add unit tests for the `active-repositories` cabal configuration field:

- `organizeByRepos`: ordering and strategy assignment with `:rest`, named repos, and error cases
- `filterSkippedActiveRepos`: filtering of skipped entries in the absence of `:rest`
- `CombineStrategy` index-combining logic (Skip/Merge/Override)
- Parse/pretty roundtrip for `ActiveRepos` (QuickCheck)
```

**Comment by geekosaur** [comment]
Do we know why this is being repeatedly queued without a merge label? The Mergify summary doesn't show it matching any of the merge rules (correct), and even if it were the presence of unresolved comments and lack of reviews should be blocking it from entering the queue at all.

**Comment by geekosaur** [comment]
@mergifyio refresh

**Comment by erikd** [comment]
I will rebase against master and address @andreabedini 's comments on Tuesday, April 7th

File: cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs (hunk 1 above)
Comment (erikd) [reply cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs:128]: The full test case is:
```
   pkgs [(repoFoo12, CombineStrategyMerge), (repoFoo2, CombineStrategyOverride)] @?= [foo2]
```
So the test is to ensure that `CombineStrategyOverride` does in fact completely override `CombineStrategyMerge`.

Am I missing something?

Comment (andreabedini) [reply cabal-install/tests/UnitTests/Distribution/Client/IndexUtils.hs:128]: I was just referring to the comment `repoFoo12 has foo-1.0 and foo-1.1`. It seems to me that repoFoo12 has foo-1.0 and foo-2.0. Maybe I misunderstood.

**Review by andreasabel — COMMENTED** [review]
I am afraid I do not even know what the `active-repositories` feature is, so I cannot judge this PR.
However, when it comes to tests: the more, the merrier!

**Comment by erikd** [comment]
The documentation for this feature is at: https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#cfg-field-active-repositories

**Review by angerman — APPROVED** [review]
The applyStrategy refactor is much more readable imo, great!

Thank you @erikd for working on this!

**Review by geekosaur — APPROVED** [review]

**Comment by angerman** [comment]
@mergifyio queue
